### ISSUE TITLE
Another crashtest

### ISSUE THREAD

**Issue by marter001** [Beep6581/RawTherapee#6475]
**Short description**
I know, it is not fair: no edit will contain more than 10 local spots, but since ist possible, i tried to have all LA tools in one list. 
Then i suddenly got a crash, which is clearly reproducable on most raw files:

**Steps to reproduce**
Open a raw file (i used cr2 or arw) in editor
Apply attached sidecar
Goto Local adjustments
Activate the spot (Dynamik Belichtung) (top down in the pic)
expand „mask and modifications“
choose the drop down „show mask“
RT will crash at this point

**how the crash will not happen:**
Open a raw file (i used cr2 or arw) in editor
Apply attached sidecar
Goto Local adjustments
Uncheck „enable spot“ (Kontrast u. wavelets)
Uncheck „enable spot“ (Neuer Spot)
Activate the spot (Dynamik Belichtung) (top down in the pic)
expand „mask and modifications“
choose the drop down „show mask“
the mask will be shown

**Additional information**
I was not able to reproduce this by creating a similar situation in a new pp3 file.

Version: 5.8-3091-gfeada4cd0
Branch: dev
Commit: feada4cd0
Commit date: 2022-04-28
Compiler: cc 11.3.0
Processor: generic x86
System: Windows
Bit depth: 64 bits
Gtkmm: V3.24.5
Lensfun: V0.3.3.0
Build type: release
Build flags:  -std=c++11 -fno-tree-loop-vectorize -mtune=generic -Werror=unused-label -Werror=delete-incomplete -fno-math-errno -Wno-attributes -Wall -Wuninitialized -Wcast-qual -Wno-deprecated-declarations -Wno-unused-result -Wunused-macros -fopenmp -Werror=unknown-pragmas -O3 -DNDEBUG -ftree-vectorize
Link flags:  -mtune=generic
OpenMP support: ON
MMAP support: ON
Build OS: MINGW64_NT-10.0-20348 3.3.4-341.x86_64 x86_64
Build date: Thu, 28 Apr 2022 06:02:37 +0000 UTC
Build epoch: 1651125757
Build UUID: 85be1867-6a6b-4912-92d7-389a34703721

[crashtest.pp3.txt](https://github.com/Beep6581/RawTherapee/files/8646554/crashtest.pp3.txt)

**Comment by Lawrence37** [comment]
Crash confirmed. Here's the backtrace.
```
#0  LUT<float>::operator[]<float, float, void>(float) const (this=0x555576bf9b60, index=-nan(0x400000)) at RawTherapee/rtengine/LUT.h:481
        idx = -2147483648
        diff = -nan(0x400000)
        p1 = -2.37460185e+27
        p2 = 4.59163468e-41
#1  0x00005555560e9789 in rtengine::ImProcFunctions::_ZN8rtengine15ImProcFunctions7wavcontERKNS_12local_paramsEPPfRNS_21wavelet_decompositionEiiRKNS_11LocwavCurveEbSA_bSA_bfiffff._omp_fn.1(void) () at RawTherapee/rtengine/iplocallab.cc:9745
        x = 96
        j = 3394
        y = 34
        lutFactorv = {inf, inf, inf, inf}
        W_L = 97
        H_L = 66
        templevel = {width = 97, rows = std::vector of length 66, capacity 66 = {0x5555835e4000, 0x5555835e4184, 0x5555835e4308, 0x5555835e448c, 0x5555835e4610, 0x5555835e4794, 0x5555835e4918, 0x5555835e4a9c, 0x5555835e4c20, 0x5555835e4da4, 0x5555835e4f28, 0x5555835e50ac, 0x5555835e5230, 0x5555835e53b4, 0x5555835e5538, 0x5555835e56bc, 0x5555835e5840, 0x5555835e59c4, 0x5555835e5b48, 0x5555835e5ccc, 0x5555835e5e50, 0x5555835e5fd4, 0x5555835e6158, 0x5555835e62dc, 0x5555835e6460, 0x5555835e65e4, 0x5555835e6768, 0x5555835e68ec, 0x5555835e6a70, 0x5555835e6bf4, 0x5555835e6d78, 0x5555835e6efc, 0x5555835e7080, 0x5555835e7204, 0x5555835e7388, 0x5555835e750c, 0x5555835e7690, 0x5555835e7814, 0x5555835e7998, 0x5555835e7b1c, 0x5555835e7ca0, 0x5555835e7e24, 0x5555835e7fa8, 0x5555835e812c, 0x5555835e82b0, 0x5555835e8434, 0x5555835e85b8, 0x5555835e873c, 0x5555835e88c0, 0x5555835e8a44, 0x5555835e8bc8, 0x5555835e8d4c, 0x5555835e8ed0, 0x5555835e9054, 0x5555835e91d8, 0x5555835e935c, 0x5555835e94e0, 0x5555835e9664, 0x5555835e97e8, 0x5555835e996c, 0x5555835e9af0, 0x5555835e9c74, 0x5555835e9df8, 0x5555835e9f7c, 0x5555835ea100, 0x5555835ea284}, buffer = std::vector of length 6402, capacity 6402 = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...}}
        WavL = 0x5555794ea000
        lutFactor = inf
        meaLut = Python Exception <class 'gdb.error'>: Request for member '_M_head_impl' is ambiguous in type 'std::tuple<LUT<float>*, std::default_delete<LUT<float> > >'. Candidates are:
  'std::default_delete<LUT<float> > std::_Head_base<1, std::default_delete<LUT<float> >, true>::_M_head_impl' (std::tuple<LUT<float>*, std::default_delete<LUT<float> > > -> std::_Tuple_impl<0, LUT<float>*, std::default_delete<LUT<float> > > -> std::_Tuple_impl<1, std::default_delete<LUT<float> > > -> std::_Head_base<1, std::default_delete<LUT<float> >, true>)
  '<unnamed type> std::_Head_base<0, LUT<float>*, false>::_M_head_impl' (std::tuple<LUT<float>*, std::default_delete<LUT<float> > > -> std::_Tuple_impl<0, LUT<float>*, std::default_delete<LUT<float> > > -> std::_Head_base<0, LUT<float>*, false>)
{_M_t = {<std::__uniq_ptr_impl<LUT<float>, std::default_delete<LUT<float> > >> = {_M_t = std::tuple containing = {[1] = 0x555576bf9b60, [2] = {<No data fields>}}}, <No data fields>}}
        WavL = 0x555555f2ef16 <LUT<float>::operator[](float __vector(4)) const+630>
        offs = 0.699999988
        mea = {-2.7062698e+12, -nan(0x400000), 1.94312516e+33, 3.0611365e-41, -0, -nan(0x400000), -2.37460185e+27, 4.59163468e-41, -2.70632013e+12, 4.59163468e-41}
        meaLut = Python Exception <class 'gdb.error'>: Request for member '_M_head_impl' is ambiguous in type 'std::tuple<LUT<float>*, std::default_delete<LUT<float> > >'. Candidates are:
  'std::default_delete<LUT<float> > std::_Head_base<1, std::default_delete<LUT<float> >, true>::_M_head_impl' (std::tuple<LUT<float>*, std::default_delete<LUT<float> > > -> std::_Tuple_impl<0, LUT<float>*, std::default_delete<LUT<float> > > -> std::_Tuple_impl<1, std::default_delete<LUT<float> > > -> std::_Head_base<1, std::default_delete<LUT<float> >, true>)
  '<unnamed type> std::_Head_base<0, LUT<float>*, false>::_M_head_impl' (std::tuple<LUT<float>*, std::default_delete<LUT<float> > > -> std::_Tuple_impl<0, LUT<float>*, std::default_delete<LUT<float> > > -> std::_Head_base<0, LUT<float>*, false>)
{_M_t = {<std::__uniq_ptr_impl<LUT<float>, std::default_delete<LUT<float> > >> = {_M_t = std::tuple containing = {[1] = 0x0, [2] = {<No data fields>}}}, <No data fields>}}
        klev = 0.699999988
        effect = 0.699999988
        lutFactor = 0
        inVals = {0, 0, 0, 0, 0, -nan(0x400000), -nan(0x400000), -nan(0x400000), -2.70629077e+12, 4.59163468e-41, -2.37460893e+27}
        src = Python Exception <class 'gdb.error'>: value requires 3157023794788302952 bytes, which is more than max-value-size
#2  0x00007ffff57c5926 in gomp_thread_start (xdata=<optimized out>) at /usr/src/debug/gcc/libgomp/team.c:125
        team = 0x555565b6e900
        task = 0x555565b6eff0
        data = <optimized out>
        pool = 0x5555665f7680
        local_fn = 0x5555560e9604 <rtengine::ImProcFunctions::_ZN8rtengine15ImProcFunctions7wavcontERKNS_12local_paramsEPPfRNS_21wavelet_decompositionEiiRKNS_11LocwavCurveEbSA_bSA_bfiffff._omp_fn.1(void)>
        local_data = 0x7fffecf58760
#3  0x00007ffff53035c2 in start_thread () at /usr/lib/libc.so.6
#4  0x00007ffff5388584 in clone () at /usr/lib/libc.so.6
```

Somehow `lutFactor` becomes infinity in `ImProcFunctions::wavcont()` (`rtengine/iplocallab.cc`). @Desmis any ideas?

**Comment by Desmis** [comment]
@Lawrence37 

I unfortunately could not take care of this problem.
I am hospitalized today, as I have announced several times my health is not good.
Maybe I will be available during the week.

@heckflosse Could you have a look, it seems perhaps an OMP problem ?
Thank you

Jacques

**Comment by Lawrence37** [comment]
@Desmis I'm sorry to hear that. Your health is more important than this project. Take care.

**Comment by chaimav** [comment]
> @Lawrence37
> 
> I unfortunately could not take care of this problem. I am hospitalized today, as I have announced several times my health is not good. Maybe I will be available during the week.
> 
> @heckflosse Could you have a look, it seems perhaps an OMP problem ? Thank you
> 
> Jacques

Our thoughts and prayers are with you

**Comment by marter001** [comment]
@Desmis i am sorry to hear this, all the best for you. Hope, you will be back healthy soon.

**Comment by heckflosse** [comment]
I think, I found the culprit. Patch follows...
Hmm, have to think a bit more about it...

**Comment by heckflosse** [comment]
This fixes the crash

```diff
diff --git a/rtengine/ipwavelet.cc b/rtengine/ipwavelet.cc
index 19ff01d92..2ac5b9b78 100644
--- a/rtengine/ipwavelet.cc
+++ b/rtengine/ipwavelet.cc
@@ -209,7 +209,7 @@ std::unique_ptr<LUTf> ImProcFunctions::buildMeaLut(const float inVals[11], const
             }
         }
     }
-    lutFactor = 1.f / lutDiff;
+    lutFactor = lutDiff == 0.f ? 0.f : 1.f / lutDiff;
     return std::unique_ptr<LUTf>(new LUTf(lutVals));
 }
 
```

**Comment by Desmis** [comment]
@heckflosse 
I tested your patch, works fine, thank you :)

But, I did a review of how masks work, especially when for the same Rt-spot, you open several tools with masks. In some cases, the behavior was not good... When you activate a mask in a tool "X", the mask in the tool "Y" must be at "Show modified image", which was not always the case. 
The "lamask_improv" branch should solve these cases.
I will soon open a pull-request (PR) to facilitate sharing

I thank you for your testimony on my health, it touches me a lot. In a few days I will start a treatment of several months which will be very tiring and restrictive. I will see how much time I have available.

jacques
